draft-ietf-jsonpath-base.md

```diff
@@ -1036,28 +1036,30 @@ The current item is selected if and only if the result is `true`.
 
 
 ~~~~ abnf
+boolean-expr     = logical-or-expr / logical-and-expr
+logical-or-expr  = logical-and-expr "||" boolean-expr ; disjunction
+                                                      ; binds less tightly than conjunction
+logical-and-expr = basic-filter /
+                   basic-filter "&&" logical-and-expr ; conjunction
+                                                      ; binds more tightly than disjunction
 
-
-boolean-expr = logical-expr
-logical-expr = ([neg-op] primary-expr) / logical-or-expr
-neg-op       = "!"                                  ; not operator
-primary-expr = "(" logical-or-expr ")"
-logical-or-expr = logical-and-expr *["||" logical-and-expr]
-logical-and-expr = comp-expr *["&&" comp-expr]
+basic-filter = comp-expr / paren-expr / ([neg-op] paren-expr)
+paren-expr   = "(" boolean-expr ")"                   ; parenthesized expression
+neg-op       = "!"                                    ; not operator
 
```

Comment:
We could select
```json
[?(@.audio && !@.video)]
```
using implicit existance test. Requiring parentheses here, i.e. `!(@.video)` would be tedious.

This brings up another issue ... see below ...

Reply:
Good point. I'd like to avoid confusing possibilities such as:
```
[?(!@.foo == bar)]
```
but I agree that it shouldn't be necessary to wrap existence terms in parentheses merely in order to negate them.

Reply:
Hmm ... if we agree to existence tests and assume 

```
{ "zero": 0, "five": 5 }
``` 

then unlike in C-like languages.

`@.five` evaluates to `true`, but consistently `!@.five` to `false` and `!!@.five` to true ... and ...
`@.zero` evaluates to `true`, `!@.zero` to `false` and `!!@.zero` to true.

Testing for values needs always to be explicit, as in `@.five == 5`, `@.five > 0`, `@.five != 6` or `@.zero != 0`.

Reply:
So then how would `!@.foo == false` be interpreted?

Reply:
if `@.foo`exists then `!@.foo` results in `false` and `!@.foo == false` will give `true` ... not to be confused with  `!(@.foo == false)` which is `true` for all values of `foo` (including `undefined`) except of  `false` ...

Reply:
So in summary, if `@.foo` holds a value of `false`, then `!@.foo == false` yields true, and `!(@.foo == false)` yields false.

I think this subtlety could bear some documentation at the very least.

Reply:
So we have the alternatives ...

1. `@.foo` always indicates existence (`true / false`) if used standalone. Drawback: inconsistency with potential value.
2.  `@.foo` results in `false` if it doesn't exist, in its value otherwise. Drawback: What's about falsy values `false, 0, null`?

We need to handle the Javascript value equivalence `undefined`, which is not a valid JSON value.

Best I can see yet, is Greg's proposal slightly modified: 

For comparisons like `@.foo > 5`, always take the node value, except it does not exist, than ... hmm ... take

* `false` or
* `null`

and let the implementation do implicite type conversion.  We cannot forbid existence operations within comparative expressions, if we don't want do error handling (which is not syntactically based).

General drawback: We cannot tell, if  `!@.foo == true` is `true` because of its value `false` or inexistence. The same effect comes with  `@.foo != true`.

General result of this discussion: We cannot reliably test whether a node exists or not.

Reply:
I'm going to push a change which renders `!@.foo == false` syntactically invalid. See what you make of it.

Reply:
In fact the changes I just pushed make existence and non-existence tests incapable of being used in comparisons, which I hope avoids some of the issues above.